### PR TITLE
SyncVersionedTransitionTask conversion logic

### PR DESCRIPTION
**Pull request by hai719**

### CODE REVIEW

File: service/history/replication/raw_task_converter.go [hunk 1]
```diff
@@ -712,3 +637,323 @@ func convertGetHistoryError(
...
+		c.taskInfo.NewRunID,
+		c.workflowCache,
+		c.eventBlobCache,
+		c.executionManager,
+		c.logger,
+	)
+	if err != nil {
+		return nil, err
+	}
+	err = c.replicationCache.Update(c.taskInfo.RunID, c.targetClusterID, nil, taskVersionHistoryItems)
```
Comment (yycptt) [inline service/history/replication/raw_task_converter.go:935]: is `taskVersionHistoryItems` only up to the nextEventID this task is trying to backfill?

File: service/history/replication/raw_task_converter.go [hunk 2]
```diff
@@ -712,3 +637,323 @@ func convertGetHistoryError(
...
+	}
+	if c.progress.TaskSent(c.taskInfo.VersionedTransition) {
+		return c.generateVerifyVersionedTransitionTask()
+	}
+
+	lca, err := c.findLCAVersionHistoryItemFromCache()
+	if err != nil {
+		return nil, err
+	}
+	if c.onCurrentBranch(lastSyncedTransition) {
```
Comment (yycptt) [inline service/history/replication/raw_task_converter.go:708]: we also need to check the the versioned Transition task we are syncing here is covered by the tombstone list in mutable state. Basically the task versioned transition >= the first versioned transition in executionInfo.subStateMachineTombstoneBatches. If not, we have to sync snapshot as well.

File: service/history/replication/raw_task_converter.go [hunk 3]
```diff
@@ -712,3 +637,323 @@ func convertGetHistoryError(
...
+	startEventID int64,
+	startEventVersion int64,
+	startEventInclusive bool,
+) ([]*commonpb.DataBlob, []byte, error) {
+	namespaceName, err := c.shardContext.GetNamespaceRegistry().GetNamespaceName(namespace.ID(c.taskInfo.NamespaceID))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resp, err := c.historyEngine.GetWorkflowExecutionRawHistoryV2(c.ctx, &historyservice.GetWorkflowExecutionRawHistoryV2Request{
```
Comment (yycptt) [inline service/history/replication/raw_task_converter.go:735]: hmm does it work if we just make the persistence call to fetch raw events, like what we did in `getEventsBlob`. maybe we can even just reuse `getEventsBlob`? I could be missing something here though.

File: service/history/replication/raw_task_converter.go [hunk 4]
```diff
@@ -712,3 +637,323 @@ func convertGetHistoryError(
...
+	}
+	return &replicationspb.ReplicationTask{
+		TaskType:     enumsspb.REPLICATION_TASK_TYPE_SYNC_VERSIONED_TRANSITION_TASK,
+		SourceTaskId: c.taskInfo.TaskID,
+		Priority:     c.taskInfo.Priority,
+		Attributes: &replicationspb.ReplicationTask_SyncVersionedTransitionTaskAttributes{
+			SyncVersionedTransitionTaskAttributes: &replicationspb.SyncVersionedTransitionTaskAttributes{
+				EventBatches:  events,
+				NextPageToken: next_page_token,
+				// TODO: SyncStateMutationAttributes
```
Comment (yycptt) [inline service/history/replication/raw_task_converter.go:851]: I think we can do this and also snapshot in a separate PR. I'd imaging a lot of logic from https://github.com/temporalio/temporal/pull/6381 can be reused here.

File: service/history/replication/raw_task_converter.go (hunk 1 above)
Comment (hai719) [reply service/history/replication/raw_task_converter.go:935]: Good point. taskVersionHistoryItems will be the current versionHistoryItems containing the first event and version. Need to truncate to the task.nextEventID

File: proto/internal/temporal/server/api/replication/v1/message.proto [hunk 5]
```diff
@@ -225,6 +225,23 @@ message VerifyVersionedTransitionTaskAttributes {
     string new_run_id = 6;
 }
 
+// (-- api-linter: core::0158::response-next-page-token-field=disabled
+//     aip.dev/not-precedent: We need to do this because current implementations. --)
 message SyncVersionedTransitionTaskAttributes {
-    // TODO
+    repeated temporal.api.common.v1.DataBlob event_batches = 1;
+    NewRunInfo new_run_info = 2;
+    bytes next_page_token = 3;
```
[the remark this answers is not in the document]
Comment (hai719) [reply proto/internal/temporal/server/api/replication/v1/message.proto]: removed.

File: service/history/replication/raw_task_converter.go [hunk 6]
```diff
@@ -712,3 +637,323 @@ func convertGetHistoryError(
...
+	// 2. task versioned transition on current transition history
+	//   2.1 never synced                                        : SyncStateMutation: events [task, current]
+	//   2.2 task.VersionedTransition is found in cache          : verify [task]
+	//   2.3 last sync on current transition history             : SyncStateMutation: events (LCA(cache,current), current]
+	//   2.4 last sync not on current transition history         : SyncStateSnapshot: events (LCA(lastSync,current),current]
+
+	transitionHistory := c.mutableState.GetExecutionInfo().TransitionHistory
+	if !c.onCurrentBranch(c.taskInfo.VersionedTransition) {
+		// 1.1 task has no events and no new run
+		if c.taskInfo.FirstEventID == c.taskInfo.NextEventID && len(c.taskInfo.NewRunID) == 0 {
```
[the remark this answers is not in the document]
Comment (hai719) [reply service/history/replication/raw_task_converter.go]: updated.

File: service/history/replication/progress_cache.go [hunk 7]
```diff
@@ -193,6 +198,47 @@ func (c *progressCacheImpl) Update(
...
+
+// func (c *ReplicationProgress) GetLastEventVersion() int64 {
+// 	items := c.LastSyncedEventHistoryItems()
+// 	if len(items) == 0 {
+// 		return 0
+// 	}
+// 	return items[len(items)-1].Version
+// }
+
+func (c *ReplicationProgress) TaskSent(versionedTransition *persistencespb.VersionedTransition) bool {
```
[the remark this answers is not in the document]
Comment (hai719) [reply service/history/replication/progress_cache.go]: updated.

File: service/history/replication/raw_task_converter.go (hunk 1 above)
Comment (hai719) [reply service/history/replication/raw_task_converter.go]: updated.

File: service/history/replication/raw_task_converter.go (hunk 2 above)
Comment (hai719) [reply service/history/replication/raw_task_converter.go]: will solve this in another PR combining with the SyncState API.

File: service/history/replication/raw_task_converter.go (hunk 4 above)
Comment (hai719) [reply service/history/replication/raw_task_converter.go]: will solve this in another PR combining with the SyncState API.

File: service/history/replication/raw_task_converter.go (hunk 3 above)
Comment (hai719) [reply service/history/replication/raw_task_converter.go]: will solve this in another PR combining with the SyncState API.

File: service/history/replication/raw_task_converter.go [hunk 8]
```diff
@@ -68,6 +69,20 @@ type (
 		clientClusterName string, // Some task converter may use the client cluster name.
 		serializer serialization.Serializer,
 	) SourceTaskConverter
+
+	syncVersionedTransitionTaskConverter struct {
```
Comment (yycptt) [inline service/history/replication/raw_task_converter.go:73]: ~~nit: feel like this separate struct is not really necessary, but no strong opinion.~~ 

I guess I have ^ impression because today the SourceTaskConverter is not really doing it's job. I see how this sync task converter saves us from passing a lot of references around. We can refactor the entire source task converter later.

**Review by yycptt — APPROVED** [review]
